pkg/streamingpromql/types/hpoint_ring_buffer.go

```diff
@@ -234,58 +203,105 @@ func (b *HPointRingBuffer) Close() {
 	b.points = nil
 }
 
-// First returns the first point in this buffer.
-// It panics if the buffer is empty.
-func (b *HPointRingBuffer) First() promql.HPoint {
-	if b.size == 0 {
-		panic("Can't get first element of empty buffer")
+type HPointRingBufferView struct {
+	buffer *HPointRingBuffer
+	size   int
+}
+
+// UnsafePoints returns slices of the points in this buffer view.
+// Either or both slice could be empty.
+// Callers must not modify the values in the returned slices or return them to a pool.
```

Comment:
(nit, I know this existed previously) could be read as "callers must return them to a pool".

```suggestion
// Callers must not modify the values in the returned slices nor return them to a pool.
```